### PR TITLE
Simplify SonicCore and redesign SonicTriton

### PR DESCRIPTION
**Pull request by kpedro88**
#### PR description:

* handle different modes in `SonicCore` w/ helper `SonicDispatcher` to reduce templating
* `SonicClientBase` now handles `mode` and `allowedTries` parameters directly (`allowedTries` can still be disabled)
* introduce `TritonData` class to handle more Triton features: multiple named inputs/outputs, different input/output types, and variable dimensions (input and output classes for `TritonClient` are now maps of `TritonData` objects)
* added test setup for a PyTorch graph network (which requires support for the features listed above) using artifacts generated by @lgray
* updated documentation, and added documentation for `SonicTriton` usage
* modified `SonicTriton` LogInfo time messages slightly to account for above changes

#### PR validation:

Ran `TritonImageProducer` and `TritonGraphProducer` many many times (and `SonicCore` unit tests).

attn: @violatingcp @drankincms @nhanvtran @jeffkrupa @jmduarte @lgray @jpata

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
Could you squash the commit history ?
It doesn't seem particularly useful to keep it around after the merge.

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
well, that's embarrassing...

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

File: HeterogeneousCore/SonicTriton/interface/TritonUtils.h
Comment (makortel) [inline HeterogeneousCore/SonicTriton/interface/TritonUtils.h:20]: Just a thought, `std::string_view` argument would allow avoiding one intermediate `std::string` when the call has a string literal (same for `wrap()`).

File: HeterogeneousCore/SonicTriton/test/TritonGraphProducer.cc
Comment (makortel) [inline HeterogeneousCore/SonicTriton/test/TritonGraphProducer.cc:87]: These two seem unused.

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (makortel) [inline HeterogeneousCore/SonicCore/src/SonicClientBase.cc:14]: Typically `Configuration` category is used for configuration errors.

File: HeterogeneousCore/SonicTriton/README.md
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/README.md:27]: Right now, `TritonClient::setBatchSize()` will emit a warning in that case, and refuse to change the batch size. (`TritonClient` keeps track of the server-side value for the maximum batch size: 0 implies no batching.)

File: HeterogeneousCore/SonicTriton/interface/TritonData.h
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/interface/TritonData.h:35]: That's the reason.

File: HeterogeneousCore/SonicTriton/interface/TritonData.h
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/interface/TritonData.h:38]: Right now, this accessor is only used in the `TritonClient` constructor. It could actually be eliminated entirely.

File: HeterogeneousCore/SonicTriton/interface/TritonData.h
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/interface/TritonData.h:62]: Is this necessary if the accessor is eliminated, as suggested above?

File: HeterogeneousCore/SonicTriton/interface/TritonUtils.h
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/interface/TritonUtils.h:9]: Is it preferable to change the name of the header to `triton_utils.h` accordingly?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:160]: This function gets the results from the results map and puts them into the output map (a quirk of the Triton API). It's only used by `TritonClient` itself, so I can just move it to the protected section.

File: HeterogeneousCore/SonicCore/interface/SonicClientBase.h
Comment (kpedro88) [reply HeterogeneousCore/SonicCore/interface/SonicClientBase.h:37]: This function is intended to be called by concrete clients to initialize the `ParameterSetDescription` object. The concrete clients have the `fillPSetDescription()` function. Let me know if there's a better pattern for this.

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:67]: It keeps `ptr` alive long enough without needing to template the whole `TritonData` class on the type of the data in `ptr` (which would make the interface much uglier and more difficult to use). This seemed like the simplest approach (after considering many other, more complex approaches.)

File: HeterogeneousCore/SonicTriton/interface/TritonUtils.h
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/interface/TritonUtils.h:17]: I'm happy to rename it. Any suggestions?

**Comment by kpedro88** [comment]
Thanks for the review @makortel. I've implemented any comments without replies. Once those with replies/questions are settled, I'll implement the rest and push.

File: HeterogeneousCore/SonicTriton/interface/TritonData.h
Comment (makortel) [reply HeterogeneousCore/SonicTriton/interface/TritonData.h:62]: I don't see `data_` being accessed in any other `const` methods, so no, `propagate_const` would not add any value.

File: HeterogeneousCore/SonicTriton/interface/TritonUtils.h
Comment (makortel) [reply HeterogeneousCore/SonicTriton/interface/TritonUtils.h:9]: That could be more clear indeed.

File: HeterogeneousCore/SonicCore/interface/SonicClientBase.h
Comment (makortel) [reply HeterogeneousCore/SonicCore/interface/SonicClientBase.h:37]: You could rename the function to e.g. `fillBasePSetDescription(edm::ParameterSetDescription& desc, bool allowRetry=true)` and then call it from the concrete clients (an alternative would be to
```cpp
using Base = SonicClient<T>;
void fillPSetDescription(...) {
  edm::ParameterSetDescription descClient;
  Base::fillPSetDescription(descClient);
  ...
`}
```
in the concrete client class).

I don't feel strongly, technically this is fine as it is (can be left if the alternatives don't feel better), it's just that an established pattern would be a little bit easier to follow.

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (makortel) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:67]: Is the aim to release the `ptr` with
```cpp
  if (callback_)
    callback_();
```
?

File: HeterogeneousCore/SonicTriton/interface/TritonUtils.h
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/interface/TritonUtils.h:17]: I'll use the suggested `throwIfError()` and also change `warn()` to `warnIfError()` for consistency.

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:67]: Yes, as called in the `reset()` function.

Comment (makortel) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:67]: Unfortunately just calling a function does not destroy the lambda. For example, you are allowed to call it many times. Only when the 'callback_` gets assigned to, the old functor is destroyed (or at least that's what I saw with gcc, not sure what the standard says).

How about using std::any to hold it?
```cpp
std::any holder_;
...
holder_ = std::move(ptr);
...
holder_.reset()
```

Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:67]: Aha, I had forgotten about `std::any`. That's indeed nicer and seems to work.

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
@makortel, @fwyzard though I recognize this is a rather substantial PR, it would be great if it could get into pre3

File: HeterogeneousCore/SonicTriton/interface/triton_utils.h
Comment (makortel) [inline HeterogeneousCore/SonicTriton/interface/triton_utils.h:18]: The `std::string_view` is cheap to copy, so it can be (and, at least implicitly, appears to be recommended) passed by value 
```suggestion
  void throwIfError(const Error& err, std::string_view msg);
```

This is my last comment. @fwyzard do you have any?

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (fwyzard) [inline HeterogeneousCore/SonicCore/src/SonicClientBase.cc:63]: @makortel as far as I know the configuration description language does not support giving a list of accepted values - unless one uses a switch statement.

Something like
```suggestion
  desc.ifValue(edm::ParameterDescription<std::string>("mode", true),
      "Sync" >> edm::EmptyGroupDescription() or
      "Async" >> edm::EmptyGroupDescription() or
      "PseudoAsync" >> edm::EmptyGroupDescription());
```
should have the equivalent behaviour (the parameter is tracked and required, does not have a default value, and will accept only the values "Sync", "Async" and "PseudoAsync".

@kpedro88 I'm not asking to make this change, just airing the possibility.

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (fwyzard) [inline HeterogeneousCore/SonicCore/src/SonicClientBase.cc:65]: is there a difference between 0 or 1 tries ?

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (fwyzard) [inline HeterogeneousCore/SonicTriton/src/TritonData.cc:94]: `std::byte` instead of `uint8_t` ?

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (fwyzard) [inline HeterogeneousCore/SonicTriton/src/TritonData.cc:97]: should there be a check on the value of `contentByteSize` ?

**Comment by fwyzard** [comment]
Nothing significant from my side - just a couple of questions and curiosities.

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:65]: No.

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:94]: This has to match the Triton API, which uses `uint8_t`.

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:63]: Thanks, I think this is a useful construction, so I've added it. (I'm keeping the exception in the constructor, just in case...)

For future reference, this pattern requires a default value to be specified for the parameter.

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:97]: added.

**Comment by kpedro88** [comment]
please test

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (fwyzard) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:69]: Thanks, good to know.
Pity for the indentation that `clang-format` imposes :-(

Comment (fwyzard) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:69]: Mhm, sorry, this was meant for the `ifValue` above :-)

File: HeterogeneousCore/SonicTriton/src/TritonData.cc
Comment (fwyzard) [reply HeterogeneousCore/SonicTriton/src/TritonData.cc:94]: OK.

**Review by fwyzard — APPROVED** [review]

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:69]: Indeed, the formatting is rather ugly. Maybe the core team can come up with a simpler syntax in the case where only `EmptyGroupDescription` is used.

**Comment by fwyzard** [comment]
+1

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (makortel) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:69]: > Maybe the core team can come up with a simpler syntax in the case where only `EmptyGroupDescription` is used.

Please open an issue.

Comment (fwyzard) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:69]: I was waiting for you to write that :-)

**Comment by silviodonato** [comment]
+1

File: HeterogeneousCore/SonicCore/src/SonicClientBase.cc
Comment (kpedro88) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:69]: See #30932

Comment (makortel) [reply HeterogeneousCore/SonicCore/src/SonicClientBase.cc:69]: I know I sound like a broken record :)

**Pull request merged**
